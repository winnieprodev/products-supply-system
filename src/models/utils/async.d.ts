export interface Error {
  message: string;
}

export interface Async<T> {
  isLoading: boolean;
  error: Error | undefined;
  data: T | undefined;
}
