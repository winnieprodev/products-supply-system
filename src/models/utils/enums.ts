export enum WrapperMode {
  full = "full",
  singleLine = "singleLine",
  tiny = "tiny",
}
