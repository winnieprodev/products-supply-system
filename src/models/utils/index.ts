export * from "./async.d";
export * from "./enums";
