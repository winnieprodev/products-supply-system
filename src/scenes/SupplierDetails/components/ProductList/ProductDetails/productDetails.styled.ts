import { GenericProductIcon } from "components";
import styled from "styled-components";
import { secondaryColorLighter } from "styles/colors";

export const Block = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;
  background: #f4f4f4;
  position: relative;
  padding: 2px;
  border-radius: 10px;
  margin-bottom: 20px;
`;

export const Content = styled.div`
  width: 375px;
  background: #f4f4f4;
  position: relative;
  align-items: center;
  justify-content: center;
`;

export const Image = styled.img`
  width: 100%;
  max-height: 100%;
  mix-blend-mode: darken;
`;

export const NoImage = styled(GenericProductIcon)`
  font-size: 100px;
  text-align: center;
  display: inherit;
  margin: 0 auto;
  padding: 4vh 0;
  opacity: 0.2;
`;

export const InfoSection = styled.div`
  padding: 15px;
  width: calc(100% - 375px);
  position: relative;
  background: #ffffff;
  border-radius: 10px;
`;

export const CloseIcon = styled.div`
  position: absolute;
  top: 12px;
  right: 16px;
  font-size: 24px;
  width: 24px;
  text-align: center;
  padding: 8px;
  border-radius: 50%;
  color: #000000;

  &:hover {
    background: ${secondaryColorLighter};
  }
`;

export const ProductTitle = styled.p`
  padding-left: 15px;
  color: #000000;
  font-size: 24px;
  font-weight: bold;
`;

export const SupplierSection = styled.div`
  padding-left: 15px;
  color: #000000;
  font-size: 16px;
`;

export const SpecialGroup = styled.div`
  display: flex;
  align-items: center;
  padding: 15px;
`;

export const SpecialText = styled.span`
  padding-left: 5px;
  padding-right: 5px;
  color: #484848;
  font-size: 14px;
`;

export const AlertIcon = styled.div`
  font-size: 15px;
  margin-right: 2px;
  color: #000000;
  opacity: 0.5;
`;

export const AmountSection = styled.div`
  padding: 10px 15px;
  align-items: center;
`;

export const AmountLabel = styled.div`
  color: #000000;
  font-size: 14px;
  padding: 15px;
`;
