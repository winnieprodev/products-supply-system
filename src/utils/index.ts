export * from "./routeUtils";
export * from "./supplierUtils";
export * from "./testUtils";
