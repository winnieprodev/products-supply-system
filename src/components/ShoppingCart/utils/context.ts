import { Async, Cart, ProductCartInfo } from "models";
import React from "react";
import { mockAsync, mockCart } from "utils";

interface ShoppingCartContextProps {
  cartId: Async<number>;
  cart: Async<Cart> | undefined;
  setCart: (cart: Async<Cart> | undefined) => void;
  setProduct: (product: ProductCartInfo) => void;
  removeProduct: (productId: string) => void;
}

export const EMPTY_CART = { products: [] };
const mockFunction = () => {};

export const ShoppingCartContext = React.createContext<ShoppingCartContextProps>(
  {
    cartId: mockAsync(0),
    cart: mockAsync(mockCart(0)),
    setCart: mockFunction,
    setProduct: mockFunction,
    removeProduct: mockFunction,
  },
);
