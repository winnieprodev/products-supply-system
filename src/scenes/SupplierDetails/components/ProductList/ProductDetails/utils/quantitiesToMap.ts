import { Map } from "immutable";
import { CartQuantity, QuantityType } from "models";

export const quantitiesToMap = (quantities: CartQuantity[]) => {
  return Map<QuantityType, number>(
    quantities.map(({ type, amount }) => [type, amount] as const) as Iterable<
      [QuantityType, number]
    >,
  );
};
