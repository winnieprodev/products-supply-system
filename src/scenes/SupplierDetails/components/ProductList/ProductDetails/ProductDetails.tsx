import {
  AddToOrderButton,
  AmountInput,
  CloseIcon,
  WarningIcon,
} from "components";
import { ShoppingCartContext } from "components/ShoppingCart";
import { SupplierLink } from "components/SupplierLink/SupplierLink";
import { Map } from "immutable";
import { CartQuantity, Product, QuantityType, Supplier } from "models";
import { useContext, useState } from "react";
import { useParams } from "react-router";
import { SupplierDetailsParams } from "routes";

import * as Styled from "./productDetails.styled";
import { quantitiesToMap } from "./utils";

interface OwnProps {
  onClose: () => void;
  item: Product;
  cartQuantities: CartQuantity[] | undefined;
  supplier: Supplier;
}

const ProductDetails: React.FC<React.PropsWithChildren<OwnProps>> = ({
  onClose,
  supplier,
  cartQuantities = [],
  item,
}) => {
  const { setProduct } = useContext(ShoppingCartContext);

  const { outletId, supplierType } = useParams<SupplierDetailsParams>();

  const [quantities, setQuantities] = useState<Map<QuantityType, number>>(
    quantitiesToMap(cartQuantities),
  );

  const getQuantityFor = (type: QuantityType) => quantities.get(type, 0);
  const setQuantityFor = (type: QuantityType) => (amount: number) => {
    setQuantities(quantities.set(type, amount));
  };

  const totalAmount = quantities.reduce((acc, value) => acc + value, 0);

  const addToOrder = () => {
    setProduct({
      productId: item.id,
      quantities: quantities.reduce<CartQuantity[]>(
        (acc, amount, type) => [...acc, { amount, type }],
        [],
      ),
    });
  };

  const isEditMode = cartQuantities.length > 0;

  return (
    <Styled.Block>
      <Styled.Content>
        {item.image ? <Styled.Image src={item.image} /> : <Styled.NoImage />}
      </Styled.Content>
      <Styled.InfoSection>
        <Styled.CloseIcon onClick={onClose}>
          <CloseIcon />
        </Styled.CloseIcon>
        <Styled.ProductTitle>{item.title}</Styled.ProductTitle>
        <Styled.SupplierSection>
          <SupplierLink
            item={supplier}
            outletId={outletId}
            supplierType={supplierType}
          />
        </Styled.SupplierSection>
        <Styled.SpecialGroup>
          {supplier.remarks && (
            <>
              <Styled.AlertIcon>
                <WarningIcon />
              </Styled.AlertIcon>
              <Styled.SpecialText>
                Special remarks: {supplier.remarks}
              </Styled.SpecialText>
            </>
          )}
        </Styled.SpecialGroup>
        {item.quantities.map(({ type, name, description }, index) => (
          <Styled.AmountSection key={index}>
            {type}:<br />
            <AmountInput
              value={getQuantityFor(type)}
              onChange={setQuantityFor(type)}
            >
              <b>{name}</b> ({description})
            </AmountInput>
          </Styled.AmountSection>
        ))}
        <AddToOrderButton
          edit={isEditMode}
          disabled={isEditMode === false && totalAmount === 0}
          onClick={addToOrder}
        />
        {item.delivery && (
          <Styled.SupplierSection>
            delivery on: {item.delivery}
          </Styled.SupplierSection>
        )}
      </Styled.InfoSection>
    </Styled.Block>
  );
};

export default ProductDetails;
