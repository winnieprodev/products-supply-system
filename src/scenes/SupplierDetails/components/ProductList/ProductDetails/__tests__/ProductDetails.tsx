import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { mockProduct, mockSupplier } from "utils";

import ProductDetails from "../ProductDetails";

describe("ProductDetails", () => {
  describe("Snapshot tests", () => {
    test("it matches the snapshot", () => {
      const { container } = render(
        <BrowserRouter>
          <ProductDetails
            item={mockProduct("id1")}
            onClose={jest.fn()}
            supplier={mockSupplier("test supplier")}
          ></ProductDetails>
        </BrowserRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
