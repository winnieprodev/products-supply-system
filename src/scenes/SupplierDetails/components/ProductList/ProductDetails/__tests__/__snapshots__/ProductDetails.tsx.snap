// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails Snapshot tests it matches the snapshot 1`] = `
.c16 {
  margin: 15px;
  width: 220px;
  display: inline-block;
}

.c10 {
  outline: none;
  cursor: pointer;
  border: 0 none;
  -webkit-transition: background ease 0.3s;
  transition: background ease 0.3s;
  background-color: #F25C05;
  color: #FFFFFF;
  opacity: 0.3;
}

.c10:hover {
  background-color: #DD5202;
}

.c10:active {
  background-color: #FF7626;
}

.c14 {
  outline: none;
  cursor: pointer;
  border: 0 none;
  -webkit-transition: background ease 0.3s;
  transition: background ease 0.3s;
  background-color: #F25C05;
  color: #FFFFFF;
}

.c14:hover {
  background-color: #DD5202;
}

.c14:active {
  background-color: #FF7626;
}

.c17 {
  border: 0;
  border-radius: 8px;
  padding: 8px 16px;
  margin: 4px;
  font-weight: 700;
  width: 100%;
  margin: 0;
}

.c11 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 22px;
  height: 22px;
  font-size: 10px;
  border-radius: 50%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c12 {
  border: 0;
  font-size: 12px;
  background-color: #EEEEEE;
  color: #5F5F5F;
  border-radius: 6px;
  padding: 8px 12px;
  outline: none;
  box-shadow: inset 1px 2px 8px rgb(0,0,0,0.05);
  -webkit-transition: background ease 0.3s;
  transition: background ease 0.3s;
  width: 40px;
  text-align: right;
}

.c12::-webkit-input-placeholder {
  color: #bdbdbd;
}

.c12::-moz-placeholder {
  color: #bdbdbd;
}

.c12:-ms-input-placeholder {
  color: #bdbdbd;
}

.c12::placeholder {
  color: #bdbdbd;
}

.c12:hover {
  box-shadow: inset 2px 3px 8px rgb(0,0,0,0.1);
  background-color: #F3F3F3;
}

.c12:active {
  background-color: #E7E7E7;
}

.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c15 {
  padding-left: 8px;
}

.c13 {
  margin: 0 2px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background: #f4f4f4;
  position: relative;
  padding: 2px;
  border-radius: 10px;
  margin-bottom: 20px;
}

.c1 {
  width: 375px;
  background: #f4f4f4;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  width: 100%;
  max-height: 100%;
  mix-blend-mode: darken;
}

.c3 {
  padding: 15px;
  width: calc(100% - 375px);
  position: relative;
  background: #ffffff;
  border-radius: 10px;
}

.c4 {
  position: absolute;
  top: 12px;
  right: 16px;
  font-size: 24px;
  width: 24px;
  text-align: center;
  padding: 8px;
  border-radius: 50%;
  color: #000000;
}

.c4:hover {
  background: #F3F3F3;
}

.c5 {
  padding-left: 15px;
  color: #000000;
  font-size: 24px;
  font-weight: bold;
}

.c6 {
  padding-left: 15px;
  color: #000000;
  font-size: 16px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px;
}

.c8 {
  padding: 10px 15px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      class="c2"
      src="src"
    />
  </div>
  <div
    class="c3"
  >
    <div
      class="c4"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-times fa-w-11 "
        data-icon="times"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
        />
      </svg>
    </div>
    <p
      class="c5"
    >
      Whipping Cream 35% 1L
    </p>
    <div
      class="c6"
    >
      <div>
        Supplier:
         
        <a
          href="/browse/undefined/undefined/test-supplier/"
        >
          test supplier
        </a>
      </div>
    </div>
    <div
      class="c7"
    />
    <div
      class="c8"
    >
      loose
      :
      <br />
      <div
        class="c9"
      >
        <button
          class="c10 c11"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-minus fa-w-14 "
            data-icon="minus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
        </button>
        <input
          class="c12 c13"
          maxlength="4"
          placeholder="0"
          value=""
        />
        <button
          class="c14 c11"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-w-14 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="c15"
        >
          <b>
            Litter
          </b>
           (
          12 x 1L
          )
        </div>
      </div>
    </div>
    <div
      class="c16"
    >
      <button
        class="c10 c17"
        disabled=""
      >
        Add to order
      </button>
    </div>
    <div
      class="c6"
    >
      delivery on: 
      tomorrow
    </div>
  </div>
</div>
`;
